cub/cub/agent/agent_segmented_scan.cuh

```diff
@@ -0,0 +1,296 @@
+ * Tuning policy types
+ ******************************************************************************/
+
+/**
+ * @brief Parameterizable tuning policy type for AgentSegmentedScan
+ *
+ * @tparam NOMINAL_BLOCK_THREADS_4B
+ *   Threads per thread block
+ *
+ * @tparam NOMINAL_ITEMS_PER_THREAD_4B
+ *   Items per thread (per tile of input)
+ *
+ * @tparam ComputeT
+ *   Dominant compute type
+ *
+ * @tparam _LOAD_ALGORITHM
+ *   The BlockLoad algorithm to use
+ *
+ * @tparam _LOAD_MODIFIER
+ *   Cache load modifier for reading input elements
+ *
+ * @tparam _STORE_ALGORITHM
+ *   The BlockStore algorithm to use
+ *
+ * @tparam _SCAN_ALGORITHM
+ *   The BlockScan algorithm to use
+ */
+template <
+  int NOMINAL_BLOCK_THREADS_4B,
+  int NOMINAL_ITEMS_PER_THREAD_4B,
+  typename ComputeT,
+  BlockLoadAlgorithm _LOAD_ALGORITHM,
+  CacheLoadModifier _LOAD_MODIFIER,
+  BlockStoreAlgorithm _STORE_ALGORITHM,
+  BlockScanAlgorithm _SCAN_ALGORITHM,
+  typename ScalingType = detail::MemBoundScaling<NOMINAL_BLOCK_THREADS_4B, NOMINAL_ITEMS_PER_THREAD_4B, ComputeT>>
+struct AgentSegmentedScanPolicy : ScalingType
+{
+  static constexpr BlockLoadAlgorithm LOAD_ALGORITHM   = _LOAD_ALGORITHM;
+  static constexpr CacheLoadModifier LOAD_MODIFIER     = _LOAD_MODIFIER;
+  static constexpr BlockStoreAlgorithm STORE_ALGORITHM = _STORE_ALGORITHM;
+  static constexpr BlockScanAlgorithm SCAN_ALGORITHM   = _SCAN_ALGORITHM;
+};
+
+/******************************************************************************
+ * Thread block abstractions
+ ******************************************************************************/
+
+namespace detail::segmented_scan
+{
+
+/**
+ * @brief AgentSegmentedScan implements a stateful abstraction of CUDA thread blocks for
+ *        participating in device-wide segmented prefix scan.
+ * @tparam AgentSegmentedScanPolicyT
+ *   Parameterized AgentSegmentedScanPolicyT tuning policy type
+ *
+ * @tparam InputIteratorT
+ *   Random-access input iterator type
+ *
+ * @tparam OutputIteratorT
+ *   Random-access output iterator type
+ *
+ * @tparam OffsetT
+ *   Signed integer type for global offsets
+ *
+ * @tparam ScanOpT
+ *   Scan functor type
+ *
+ * @tparam InitValueT
+ *   The init_value element for ScanOpT type (cub::NullType for inclusive scan)
+ *
+ * @tparam AccumT
+ *   The type of intermediate accumulator (according to P2322R6)
+ */
+template <typename AgentSegmentedScanPolicyT,
+          typename InputIteratorT,
+          typename OutputIteratorT,
+          typename OffsetT,
+          typename ScanOpT,
+          typename InitValueT,
+          typename AccumT,
+          bool ForceInclusive = false>
+struct AgentSegmentedScan
+{
+  //---------------------------------------------------------------------
+  // Types and constants
+  //---------------------------------------------------------------------
+
+  // The input value type
+  using InputT = cub::detail::it_value_t<InputIteratorT>;
```

Comment:
Our [style guide](https://github.com/NVIDIA/cccl/wiki/Cpp-Coding-Guidelines#cub-and-thrust) is still evolving and needs to be documented more prominently. Entities that user's will not or cannot interface with are `snake_case`: that is, type aliases, variables, etc. Note, template parameters are `CamelCase`.

```suggestion
  using input_t = cub::detail::it_value_t<InputIteratorT>;
```